c10/test/util/complex_test_common.h

```diff
@@ -389,6 +389,32 @@ MAYBE_GLOBAL void test_arithmetic() {
   test_arithmetic_<double>();
 }
 
+template<typename T, typename int_t>
+void test_binary_ops_for_int_type_(T real, T img, int_t num) {
+  c10::complex<T> c(real, img);
+  ASSERT_EQ(c + num, c10::complex<T>(real + num, img));
+  ASSERT_EQ(num + c, c10::complex<T>(num + real, img));
+  ASSERT_EQ(c - num, c10::complex<T>(real - num, img));
+  ASSERT_EQ(num - c, c10::complex<T>(num - real, -img));
+  ASSERT_EQ(c * num, c10::complex<T>(real * num, img * num));
+  ASSERT_EQ(num * c, c10::complex<T>(num * real, num * img));
+  ASSERT_EQ(c / num, c10::complex<T>(real / num, img / num));
+  ASSERT_EQ(num / c, c10::complex<T>(num * real / std::norm(c), -num * img / std::norm(c)));
+}
+
+template<typename T>
+void test_binary_ops_for_all_int_types_(T real, T img, int8_t i) {
+  test_binary_ops_for_int_type_<T, int8_t>(real, img, i);
+  test_binary_ops_for_int_type_<T, int16_t>(real, img, i);
+  test_binary_ops_for_int_type_<T, int32_t>(real, img, i);
+  test_binary_ops_for_int_type_<T, int64_t>(real, img, i);
+}
+
+TEST(TestArithmeticIntScalar, All) {
```

Comment:
@zasdfgbnm I ended up using the `TEST` macro to avoid the Windows CI error

Reply:
That is OK, could you add a comment?

Reply:
Probably make a separate PR and add some Windows CUDA comments somewhere? I'm not sure what the true reason this is (but I'm aware that Windows CUDA has a lot of stern requirements). Perhaps @ezyang knows more about this? Thanks!

The Windows error is:

```
C:/Users/circleci/project\c10/test/util/complex_test_common.h(407): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(408): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(409): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here

C:/Users/circleci/project\c10/test/util/complex_test_common.h(410): error: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during instantiation of "void arithmetic::test_binary_ops_for_all_int_types_(T, T, int8_t) [with T=float]" 
(414): here
```

The change that makes the Windows error message go away:

```diff
diff --git a/c10/test/util/complex_test_common.h b/c10/test/util/complex_test_common.h
index 4fde8da2367e..fe61ca0e231d 100644
--- a/c10/test/util/complex_test_common.h
+++ b/c10/test/util/complex_test_common.h
@@ -390,27 +390,27 @@ MAYBE_GLOBAL void test_arithmetic() {
 }
 
 template<typename T, typename int_t>
-C10_HOST_DEVICE void test_binary_ops_for_int_type_(T real, T img, int_t num) {
+void test_binary_ops_for_int_type_(T real, T img, int_t num) {
   c10::complex<T> c(real, img);
   static_assert(c + num == c10::complex<T>(real + num, img), "");
   static_assert(num + c == c10::complex<T>(num + real, img), "");
@@ -403,7 +403,7 @@ inline void test_binary_ops_for_int_type_(T real, T img, int_t num) {
 } 

 template<typename T>
-C10_HOST_DEVICE void test_binary_ops_for_all_int_types_(T real, T img, int8_t i) {
+void test_binary_ops_for_all_int_types_(T real, T img, int8_t i) {
   test_binary_ops_for_int_type_<T, int8_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int16_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int32_t>(real, img, i);
   test_binary_ops_for_int_type_<T, int64_t>(real, img, i);
 }
 
-MAYBE_GLOBAL void test_arithmetic_int_scalar() {
+TEST(TestArithmeticIntScalar, All) {
   test_binary_ops_for_all_int_types_<float>(1.0, 0.1, 1);
   test_binary_ops_for_all_int_types_<double>(-1.3, -0.2, -2);
 }
```

Reply:
I don't know, it looks like you have declared everything as `C10_HOST_DEVICE` where it should be. And sure, feel free to do it in a separate PR.